harness/determined/core/_metrics.py

```diff
@@ -0,0 +1,168 @@
+import logging
+import queue
+import threading
+from typing import Any, Dict, List, Optional
+
+from determined.common import api
+from determined.common.api import bindings
+
+logger = logging.getLogger("determined.core")
+
+
+class MetricsContext:
+    """Gives access to metrics reporting during trial tasks.
+
+    Metrics reported to ``MetricsContext`` are published to a queue, which is consumed by a
+    background thread that reports them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        trial_id: int,
+        run_id: int,
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._error_queue = queue.Queue()
+        self._shipper = _Shipper(
+            session=self._session,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            error_queue=self._error_queue,
+        )
+
+    def report(
+        self,
+        group: str,
+        steps_completed: int,
+        metrics: Dict[str, Any],
+        batch_metrics: Optional[List[Dict[str, Any]]] = None,
+    ) -> None:
+        # Check for thread exceptions here since we're not polling.
+        if not self._error_queue.empty():
+            err_msg = self._error_queue.get(block=False)
+            logger.error(f"Error reporting metrics: {err_msg}")
+            raise err_msg
+
+        self._shipper.publish_metrics(
+            group=group,
+            steps_completed=steps_completed,
+            metrics=metrics,
+            batch_metrics=batch_metrics,
+        )
+
+    def start(self) -> None:
+        self._shipper.start()
+
+    def close(self) -> None:
+        self._shipper.stop()
+        self._shipper.join()
```

Comment:
instead of a plain join, can we do a join with a time out logic similar to https://github.com/determined-ai/determined/blob/main/harness/determined/core/_log_shipper.py#L107

way too many times we had cases when the process is stuck on bg thread join, because there's a networking hangup or something. hard to debug every time. so I'd suggest all non-daemon threads which are joined must have a timeout and logging of some sort for "it took too long and we gave up".

Reply:
good point. done, it's gated behind a check for if the inbound queue is empty. because if the core context is waiting on metrics to finish reporting, we can't just time it out. 

if the queue is not empty and there's a hang, i'd wager it's probably better to risk the hang than to risk killing some metric reports.

Reply:
if the core context is waiting on metrics to finish reporting, but they are stuck for a long time, and that join takes a while, it'd still be opaque for a user why their script doesn't exit. I'd suggest we should also join with a timeout in a loop, and print a log message (after 10 seconds) that we are waiting for that to finish.

Reply:
done